lib/fastlane/plugin/wpmreleasetoolkit/helper/strings_file_validation_helper.rb

```diff
@@ -0,0 +1,129 @@
+      #
+      # @param [String] file The path to the file to inspect.
+      # @return [<Hash>] A `Hash` with keys the key values read from the
+      #         `.strings` file. The value of each entry is an array with the
+      #         line numbers where the key occurs.
+      def self.get_keys_from_strings(file:)
+        keys_with_lines = {}
+
+        state = State.new(context: :root, buffer: StringIO.new, in_escaped_ctx: false, found_key: nil)
+
+        File.readlines(file).each_with_index do |line, line_no|
+          line.chars.each_with_index do |c, col_no|
+            # Handle escaped characters at a global level. This is more
+            # straightforward than having a `TRANSITIONS` table that account
+            # for it.
+            if state.in_escaped_ctx || c == '\\'
+              # Just because we check for escaped characters at the global
+              # level, it doesn't mean we allow them in every context.
+              raise "Found escaped character outside of allowed contexts (current context: #{state.context})" unless %i[in_quoted_key in_quoted_value in_block_comment].include?(state.context)
+
+              state.buffer.write(c) if state.context == :in_quoted_key
+              state.in_escaped_ctx = !state.in_escaped_ctx
+              next
+            end
+
+            # Look at the transitions table for the current context, and find
+            # the first transition matching the current character
+            (_, next_context) = TRANSITIONS[state.context].find { |regex, _| c.match?(regex) } || [nil, nil]
+            raise "Invalid character `#{c}` found on line #{line_no + 1}, col #{col_no + 1}" if next_context.nil?
+
+            state.context = next_context.is_a?(Proc) ? next_context.call(state, c) : next_context
+            next unless state.found_key
+
+            # If we just exited the :in_quoted_key context and thus have found
+            # a new key, process it
+            key = state.found_key.dup
+            state.found_key = nil
+
+            keys_with_lines[key] = keys_with_lines[key].nil? ? [line_no + 1] : keys_with_lines[key].push(line_no + 1)
+          end
+        end
+
+        keys_with_lines
+      end
+
+      # Inspects the given `.strings` file for duplicated keys, returning them
+      # if any.
+      #
+      # @param [String] file The path to the file to inspect.
+      # @return [Array<Hash>] Array of all the duplicated keys. Each entry is a
+      #         a `Hash` with keys: `key` for the key value and `lines` with an
+      #         array of line numbers at which the key occurs.
+      def self.find_duplicated_keys(file:)
+        get_keys_from_strings(file: file).reduce([]) do |accumulator, (key, lines)|
+          if lines.length >= 2
+            # Duplicate key, keep
+            accumulator.append({ key: key, lines: lines })
+          else
+            # Valid key, discard
+            accumulator
+          end
+        end
+      end
```

Comment:
I think you should be able to avoid the `reduce` complexity completely here and just call `delete_if` or `keep_if` on the hash instead:
```suggestion
      def self.find_duplicated_keys(file:)
        get_keys_from_strings(file: file).keep_if { |_, lines| lines.length > 1 }
      end
```

Reply:
Wow, cool methods 😄 

To use `keep_if`, we'd need to update the return type from `get_keys_from_strings` to be already in the `{ key:, lines: }` format. That's okay for me. 

Originally, I called `.select` then `.map`, then replaced with `reduce` to iterate only once. To be fair, though, because the `.map` would happen after the `.keep_if` or `.select`, it would only touch the limited list of results, which would make the second iteration negligible in most case. Also... this is all premature optimization 🙃 

I was thinking about this after signing off and I'm in fact a bit unsure about the two methods returning data in a different format. But, while `[{ key:, line: }]` seems good for `find_duplicated_keys(file:)` (which I now realized I should rename to `find_repeated_keys(file:)`, `Hash<key:, line:>` seemed better for the other.

🤔 Maybe I could get the best of both worlds by returning this from `get_keys_from_strings`:

```ruby
{
		foo: { key: 'foo', lines: [1] },
		bar: { key: 'bar', lines: [2] }
}
```

In `find_repeated_keys` I could then call `.values .keep_if`. What do you think?

Reply:
Tbh I'm not sure I see the point of using the `Array<Hash<key:,lines:>>` intermediate format (ie `[{key: 'foo', lines: [2,6,12]}, {key: 'bar', lines: [7,42,1337]}]`) while this is essentially an unordered mapping listing an array of lines associated with each key anyway … aka a `Hash`

So personally I'd go with `Hash<String, Array<Int>>` (ie `{'foo': [2,6,12], 'bar': [7,42,1337]}`) for both cases.

Reply:
Fair point... I thought of the keys to make the data clearer to access, but the use case and data footprint is so small that the simpler `{ foo: [2,6,12] }` doesn't cause any confusion. 👍